src/backend/distributed/utils/node_metadata.c

```diff
@@ -120,32 +150,104 @@ master_remove_node(PG_FUNCTION_ARGS)
 	text *nodeName = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
 	char *nodeNameString = text_to_cstring(nodeName);
-	bool forceRemove = false;
-	RemoveNodeFromCluster(nodeNameString, nodePort, forceRemove);
+
+	RemoveNodeFromCluster(nodeNameString, nodePort);
 
 	PG_RETURN_VOID();
 }
 
 
 /*
- * master_disable_node function removes the provided node from the pg_dist_node table of
- * the master node and all nodes with metadata regardless of the node having an active
- * shard placement.
- * The call to the master_remove_node should be done by the super user.
- * This function also deletes all reference table placements belong to the given node from
- * pg_dist_shard_placement, but it does not drop actual placement at the node. In the case
- * of re-adding the node, master_add_node first drops and re-creates the reference tables.
+ * master_disable_node function sets nodestate of the provided node to inactive
+ * at master node and all nodes with metadata regardless of the node having an
+ * active shard placement.
+ * The call to the master_disable_node should be done by the super user.
+ * This function also deletes all reference table placements belong to the given
+ * node from pg_dist_shard_placement, but it does not drop actual placement at
+ * the node. In the case of re-activating the node, master_add_node first drops
+ * and re-creates the reference tables.
  */
 Datum
 master_disable_node(PG_FUNCTION_ARGS)
 {
+	text *nodeNameText = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+
+	char *nodeName = text_to_cstring(nodeNameText);
+	bool hasShardPlacements = false;
+
+	DeleteAllReferenceTablePlacementsFromNode(nodeName, nodePort);
+	SetNodeState(nodeName, nodePort, NODE_STATE_INACTIVE);
+
+	hasShardPlacements = NodeHasActiveShardPlacements(nodeName, nodePort);
+
+	if (hasShardPlacements)
+	{
+		ereport(NOTICE, (errmsg("Node %s:%d has active shard placements. Some queries "
+								"may fail after this operation. Use "
+								"SELECT master_activate_node('%s', %d) to activate this "
+								"node back.",
+								nodeName, nodePort, nodeName, nodePort)));
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * master_activate_node UDF activates the given node. It sets the node's state
+ * as active and replicates all reference tables to that node.
+ */
+Datum
+master_activate_node(PG_FUNCTION_ARGS)
+{
 	text *nodeName = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
+
 	char *nodeNameString = text_to_cstring(nodeName);
-	bool forceRemove = true;
-	RemoveNodeFromCluster(nodeNameString, nodePort, forceRemove);
+	Datum nodeRecord = 0;
 
-	PG_RETURN_VOID();
+	nodeRecord = ActivateNode(nodeNameString, nodePort);
```

Comment:
I guess `nodeid` would be useful when we implement #968. I am not sure if they have non-replacable usage areas right now.